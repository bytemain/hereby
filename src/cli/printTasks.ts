import commandLineUsage from "command-line-usage";
import pc from "picocolors";

import type { Task } from "../index.js";
import { stringSorter, taskSorter } from "./utils.js";

export function printTasks(tasks: Task[], defaultTask?: Task) {
    tasks = tasks.slice(0).sort(taskSorter);

    const sections: commandLineUsage.Section[] = [];
    sections.push({
        header: "Available tasks",
        content: tasks.map((task) => {
            const blueName = pc.blue(task.options.name);
            const name = task !== defaultTask ? blueName : `${blueName} (default)`;

            const descriptionParts: string[] = [];
            if (task.options.description) {
                descriptionParts.push(task.options.description);
            }

            if (task.options.dependencies && task.options.dependencies.length > 0) {
                const deps = task.options.dependencies
                    .map((task) => task.options.name)
                    .sort(stringSorter)
                    .map(pc.blue);
                descriptionParts.push(`Depends on: ${deps.join(", ")}`);
            }

            return { name, description: descriptionParts.join("\n") };
        }),
    });

    console.log(commandLineUsage(sections));
}
