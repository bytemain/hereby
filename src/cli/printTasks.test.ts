process.env["FORCE_COLOR"] = "1";

import { mockConsoleLog } from "jest-mock-process";

import { task } from "../index.js";
import { printTasks } from "./printTasks.js";

test("printTasks", () => {
    const a = task({
        name: "a",
        description: "This is task a. It works pretty well.",
    });

    const b = task({
        name: "b",
        dependencies: [a],
    });

    const c = task({
        name: "c",
        description: "This is task c. ".repeat(10),
        dependencies: [a, b],
    });

    const d = task({
        name: "d",
    });

    const saveColumns = process.stdout.columns;
    const saveTty = process.stdout.isTTY;
    process.stdout.columns = 80;
    process.stdout.isTTY = false;
    const mockLog = mockConsoleLog();

    const lines: string[] = [];

    mockLog.mockImplementation((message) => lines.push(message));

    printTasks([a, c, d], d);

    mockLog.mockRestore();
    process.stdout.columns = saveColumns;
    process.stdout.isTTY = saveTty;

    const stdout = lines.join("\n");
    expect(stdout).toMatchSnapshot("stdout");
});
