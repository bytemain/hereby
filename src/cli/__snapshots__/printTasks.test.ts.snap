// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printTasks: stdout 1`] = `
"
[4m[1mAvailable tasks[22m[24m

  [34ma[39m             This is task a. It works pretty well.                           
  [34mc[39m             This is task c. This is task c. This is task c. This is task c. 
                This is task c. This is task c. This is task c. This is task    
                c. This is task c. This is task c.                              
                Depends on: [34ma[39m, [34mb[39m                                                
  [34md[39m (default)                                                                   
"
`;
